File: src/main/java/main/sorting/impl/CountingSort.java
package main.sorting.impl;

import main.sorting.Sort;

public class CountingSort implements Sort<int[]> {
	private int[] table;

	@Override
	public int[] getSortedData(int[] tab) {
		this.table = tab;
		sort();
		return table;
	}

	private void sort() {
		int minVal = table[0];
		int maxVal = table[0];

		for (int i = 1; i < table.length; i++) {
			if (table[i] < minVal) {
				minVal = table[i];
			} else if (table[i] > maxVal) {
				maxVal = table[i];
			}
		}

		int[] countTab = new int[maxVal - minVal + 1];
		for (int value : table) {
			countTab[value - minVal]++;
		}

		int tableIndex = 0;
		for (int i = 0; i < countTab.length; i++) {
			while (countTab[i] != 0) {
				table[tableIndex] = i + minVal;
				countTab[i]--;
				tableIndex++;
			}
		}
	}
}

File: src/main/java/main/sorting_kotlin/impl/CountingSort.kt
package main.sorting_kotlin.impl

import main.sorting_kotlin.Sort

class CountingSort : Sort<IntArray> {
    override fun getSorted(table: IntArray): IntArray {
        var minVal: Int = table[0]
        var maxVal: Int = table[0]

        for (item in table) {
            if (item < minVal) {
                minVal = item
            } else if (item > maxVal) {
                maxVal = item
            }
        }

        val countTab: IntArray = IntArray(maxVal - minVal + 1)
        table.forEach { item -> countTab[item - minVal]++ }

        var tableIndex: Int = 0
        for (i in 0 until countTab.size) {
            while (countTab[i] != 0) {
                table[tableIndex] = i + minVal
                countTab[i]--
                tableIndex++
            }
        }

        return table
    }
}
